Infer legacy Volumes that do not have the supervised label

20k-ultra opened the pull request:
I spent a bit of time trying to figure out what the least impactful way to make the Supervisor detect Volumes without the supervised label. This fix solves that by using existing functions namely,  `deconstructDockerName` which tries to extract Volume info such as appId from a given volume ID. If the Supervisor cannot extract info from a Volume's inspect info which it created then it can assume that volume was not created by the Supervisor. Therefore, this patch allows the Supervisor to detect such legacy volumes and also logs whenever a Volume is not supervised to help with troubleshooting in the future if the Supervisor is ignoring a volume.

Change-type: patch
Closes: #1604
Signed-off-by: Miguel Casqueira <miguel@balena.io>

20k-ultra commented:
working on tests but wanted to share patch to get the conversation going.

20k-ultra commented on src/compose/volume-manager.ts:
```diff
@@ -23,9 +24,29 @@ export async function get({ name, appId }: VolumeNameOpts): Promise<Volume> {
+		volumes = await listWithBothLabels();
+	}
+	// Normalize inspect information to Volume types and filter any that fail
+	return volumes.reduce((volumesList, volumeInfo) => {
```
@Page- what are your thoughts on a cleaner way of doing this. I never check what options lodash has so maybe there is one. To be clear, I'm happy with this solution but I think you've pointed out smaller LOC implementations that achieve the same thing. I basically need a filter catch function..

pipex commented on src/compose/volume-manager.ts:
```diff
@@ -23,9 +24,29 @@ export async function get({ name, appId }: VolumeNameOpts): Promise<Volume> {
+			const volume = Volume.fromDockerVolume(volumeInfo);
+			volumesList.push(volume);
+		} catch (e) {
+			if (e.message.substring(0, 16) === 'Could not detect') {
```
I think creating a new exception type and throw that in `Volume.deconstructDockerName` might be  a better option here @20k-ultra

pipex replied:
I think this doesn't work for correctly named volumes without the `io.balena.supervised` label, right? In that case `Volume.deconstructDockerName` doesn't throw and it is also not catched by `listByLabel`

pipex replied:
Is there any drawback to always using `list()` instead of `listByLabel`? If there isn't I would get rid of the function parameter and just use `list()` by default

20k-ultra replied:
I agree with creating a new exception type that we can use for better error handling of this type. I'll do this.

This error is only for failure to parse a volume name, there's no need for it to be aware of the `io.balena.supervised` label so your observations are by design.

I kept listByLabel (previously listWithBothLabels) even though no one is using it because when I'm patching an issue I want to avoid introducing new regressions at all costs so I write the patch in a way that fixes the issue without introducing new behavior. By using this toggle (allowLegacy) the codebase acts the same unless when this value is true. However, whenever I make a PR patch and the reviewer thinks there's optimizations we can squeeze in at the same time I'm cool to add them. i'll clean this up too.

pipex replied:
Oh, I thought those volumes without the label were causing the issue. I see now.

For  the adding the `allowLegacy` arg, if you feel it might be safer to do it that way, then let's keep it. It is true that just relying on naming to identify volumes might be easy to break, so do what you believe it's best

20k-ultra commented:
All the files where I change values like '{}' to '[]' is because they are suppose to be arrays not objects. I learned this because I added listVolumes which some functions perform array functions such as reduce. This is when i saw that the default values of volumes, networks, containers, etc being objects was incorrect. Since I fixed that in mocked-dockerode I went to the test files to make sure they also had arrays.

pipex commented:
Nice catch on the '{}', my guess is that whoever wrote those tests got confused with the target state endpoint format (where they are objects)

pipex reviewed (approved):
LGTM
